sound/soc/sof/imx/imx8.c

```diff
@@ -400,6 +402,92 @@ static int imx8_get_bar_index(struct snd_sof_dev *sdev, u32 type)
 	}
 }
 
+static void imx8_suspend(struct snd_sof_dev *sdev)
+{
+	int i;
+	struct imx8_priv *priv = (struct imx8_priv *)sdev->pdata->hw_pdata;
+
+	for (i = 0; i < DSP_MU_CHAN_NUM; i++)
+		imx_dsp_free_channel(priv->dsp_ipc, i);
+
+	imx8_disable_clocks(sdev, priv->clks);
+}
+
+static int imx8_resume(struct snd_sof_dev *sdev)
+{
+	struct imx8_priv *priv = (struct imx8_priv *)sdev->pdata->hw_pdata;
+	int ret;
+	int i;
+
+	ret = imx8_enable_clocks(sdev, priv->clks);
+	if (ret < 0)
+		return ret;
+
+	for (i = 0; i < DSP_MU_CHAN_NUM; i++)
+		imx_dsp_request_channel(priv->dsp_ipc, i);
+
+	return 0;
+}
+
+static int imx8_dsp_runtime_resume(struct snd_sof_dev *sdev)
+{
+	int ret;
+	const struct sof_dsp_power_state target_dsp_state = {
+		.state = SOF_DSP_PM_D0,
+	};
+
+	ret = imx8_resume(sdev);
+	if (ret < 0)
+		return ret;
+
+	return snd_sof_dsp_set_power_state(sdev, &target_dsp_state);
+}
+
+static int imx8_dsp_runtime_suspend(struct snd_sof_dev *sdev)
+{
+	const struct sof_dsp_power_state target_dsp_state = {
+		.state = SOF_DSP_PM_D3,
+	};
+
+	imx8_suspend(sdev);
+
+	return snd_sof_dsp_set_power_state(sdev, &target_dsp_state);
+}
+
+static int imx8_dsp_suspend(struct snd_sof_dev *sdev, unsigned int target_state)
+{
+	const struct sof_dsp_power_state target_dsp_state = {
+		.state = target_state,
+	};
+
+	if (!pm_runtime_suspended(sdev->dev))
```

Comment:
@dbaluta looks good now but just out of curiosity why do you need this check here? In the case of PCI device, a system suspend always runtime resumes the device before calling the system uspend callback. Is it not the case for OF device?

Reply:
In OF case a system suspend doesn't behave like the PCI case so the device isn't runtime resumed before system suspend.

I added that check in order to void the situation immediately after boot where the DSP is runtime suspended and then the user (or the system) initiates system suspend.

Reply:
I see I wonder if in that case, if you return a 1 in the prepare callback, you'd never receive the suspend callback. 

 "For devices supporting runtime power management, the return value of the prepare callback can be used to indicate to the PM core that it may safely leave the device in runtime suspend (if runtime-suspended already), provided that all of the device's descendants are also left in runtime suspend. Namely, if the prepare callback returns a positive number and that happens for all of the descendants of the device too, and all of them (including the device itself) are runtime-suspended, the PM core will skip the suspend, suspend_late and suspend_noirq phases as well as all of the corresponding phases of the subsequent device resume for all of these devices. In that case, the ->complete callback will be invoked directly after the ->prepare callback and is entirely responsible for putting the device into a consistent state as appropriate." from https://01.org/linuxgraphics/gfx-docs/drm/driver-api/pm/devices.html

this never worked for the PCI device for us, just wondering if this would work for you.

Reply:
I remember that I have tried all sort of tricks at that moment and all this kind of implementation worked for me. `snd_sof_prepare` always returns 0 in SOF as of now.